### PR TITLE
add functionality to access SCIP_NLROWs

### PR DESCRIPTION
**Pull request by mueldgog**
I have added a new class for SCIP_NLROWs. This is useful if one needs to have a global view on the problem. Iterating through the constraints is in general not possible because there might be some upgraded constraints that are not supported by the Python interface.

### CODE REVIEW

File: src/pyscipopt/scip.pyx [hunk 1]
```diff
@@ -407,6 +407,71 @@ cdef class Row:
...
+        def __get__(self):
+            cname = bytes( SCIPnlrowGetName(self.scip_nlrow) )
+            return cname.decode('utf-8')
+
+    def getConstant(self):
+        """returns the constant of a nonlinear row"""
+        return SCIPnlrowGetConstant(self.scip_nlrow)
+
+    def getLinearTerms(self):
+        """returns a list of linear terms of a nonlinear row"""
```
Comment (fserra) [inline src/pyscipopt/scip.pyx]: `returns a list of tuples (var, coef)` What about a dict between var.ptr and lincoef

File: src/pyscipopt/scip.pyx [hunk 2]
```diff
@@ -407,6 +407,71 @@ cdef class Row:
...
+        cdef SCIP_VAR** linvars = SCIPnlrowGetLinearVars(self.scip_nlrow)
+        cdef SCIP_Real* lincoefs = SCIPnlrowGetLinearCoefs(self.scip_nlrow)
+        cdef int nlinvars = SCIPnlrowGetNLinearVars(self.scip_nlrow)
+        return [(Variable.create(linvars[i]), lincoefs[i]) for i in range(nlinvars)]
+
+    def getQuadraticTerms(self):
+        """returns a list of quadratic terms of a nonlinear row"""
```
Comment (fserra) [inline src/pyscipopt/scip.pyx]: list of tuples var1 var2 coef

File: src/pyscipopt/scip.pyx [hunk 3]
```diff
@@ -2293,6 +2358,21 @@ cdef class Model:
...
+    def isNLPConstructed(self):
+        """returns whether SCIP's internal NLP has been constructed"""
+        return SCIPisNLPConstructed(self._scip)
+
+    def getNNlRows(self):
+        """gets current number of nonlinear rows in SCIP's internal NLP"""
+        return SCIPgetNNLPNlRows(self._scip)
+
+    def getNlRows(self):
+        """returns the nonlinear rows in SCIP's internal NLP"""
```
Comment (fserra) [inline src/pyscipopt/scip.pyx]: a list with the nonlin..

File: src/pyscipopt/scip.pyx (hunk 1 above)
Comment (mueldgog) [reply src/pyscipopt/scip.pyx]: d1edb6f

**Comment by mueldgog** [comment]
> looks good. What about expression trees?

I only care about quadratics for now. To be honest, I don't know the expression structure in PySCIPOpt good enough to implement this. I would postpone this and then implement this properly when consexpr has been released.

**Pull request merged**